src/corelibs/U2View/src/ov_msa/exclude_list/MsaExcludeList.cpp

```diff
@@ -0,0 +1,593 @@
+    connect(nameListView, &QWidget::customContextMenuRequested, this, &MsaExcludeListWidget::showNameListContextMenu);
+
+    sequenceView = new QPlainTextEdit(this);
+    sequenceView->setObjectName("exclude_list_sequence_view");
+    sequenceView->setReadOnly(true);
+    sequenceView->setTextInteractionFlags(Qt::TextSelectableByKeyboard | Qt::TextSelectableByMouse);
+
+    namesAndSequenceSplitter->addWidget(nameListView);
+    namesAndSequenceSplitter->setStretchFactor(0, 1);
+    namesAndSequenceSplitter->setCollapsible(0, false);
+    namesAndSequenceSplitter->addWidget(sequenceView);
+    namesAndSequenceSplitter->setStretchFactor(1, 3);
+    namesAndSequenceSplitter->setCollapsible(1, false);
+
+    this->setContextMenuPolicy(Qt::PreventContextMenu);
+    nameListView->setContextMenuPolicy(Qt::CustomContextMenu);
+
+    moveToMsaAction = new QAction(tr("Move to alignment"), this);
+    moveToMsaAction->setToolTip(tr("Move selected Exclude List sequences to MSA"));
+    moveToMsaAction->setIcon(QIcon(":core/images/arrow-move-up.png"));
+    connect(moveToMsaAction, &QAction::triggered, this, &MsaExcludeListWidget::moveExcludeListSelectionToMaObject);
+
+    auto moveToMsaButton = new QToolButton();
+    moveToMsaButton->setObjectName("exclude_list_move_to_msa_button");
+    moveToMsaButton->setDefaultAction(moveToMsaAction);
+    moveToMsaButton->setToolButtonStyle(Qt::ToolButtonIconOnly);
+    toolbarLayout->addWidget(moveToMsaButton);
+    toolbarLayout->addStrut(10);
+
+    auto moveFromMsaAction = viewContext->getMoveMsaSelectionToExcludeListAction(msaEditor);
+    if (moveFromMsaAction != nullptr) {
+        auto moveFromMsaButton = new QToolButton();
+        moveFromMsaButton->setObjectName("exclude_list_move_to_exclude_list_button");
+        moveFromMsaButton->setDefaultAction(moveFromMsaAction);
+        moveFromMsaButton->setToolButtonStyle(Qt::ToolButtonIconOnly);
+        toolbarLayout->addWidget(moveFromMsaButton);
+        toolbarLayout->addStrut(10);
+    }
+
+    auto msaObject = msaEditor->getMaObject();
+    connect(msaObject, &MultipleSequenceAlignmentObject::si_alignmentChanged, this, &MsaExcludeListWidget::handleUndoRedoInMsaEditor);
+
+    excludeListFilePath = msaEditor->property(PROPERTY_LAST_USED_EXCLUDE_LIST_FILE).toString();
+    if (excludeListFilePath.isEmpty() || !QFileInfo::exists(excludeListFilePath)) {
+        GUrl msaUrl = msaObject->getDocument()->getURL();
+        excludeListFilePath = msaUrl.dirPath() + "/" + msaUrl.baseFileName() + "." + EXCLUDE_LIST_FILE_SUFFIX;
+        if (GUrl(excludeListFilePath) == msaUrl) {
+            excludeListFilePath = msaUrl.getURLString() + "." + EXCLUDE_LIST_FILE_SUFFIX;
+        }
+    }
+
+    connect(AppContext::getTaskScheduler(), &TaskScheduler::si_stateChanged, this, &MsaExcludeListWidget::trackMsaObjectSaveTask);
+
+    toolbarLayout->addStretch();
+
+    loadExcludeList(true);
+}
+
+MsaExcludeListWidget::~MsaExcludeListWidget() {
```

Comment:
Destructor isn't suppose to do some extra work, it's for destruction only. This stuff should be done in a special functions [destroyed](https://doc.qt.io/qt-5/qobject.html#destroyed)

Reply:
Why do destructors exist at all in C++ in that case? 
Are they really called at all unless mentioned explicitly (like constructors)? 
A lot of QT classes have destructors (QObject/QWidget(s)/QModel(s)), a lot of std classes also have destructors.

The destructor usage may be safe or unsafe like any other function and this is a goal of any code review: to reveal these unsafe places

Reply:
Another reason why "destroyed" is not suitable here:
"destroyed" should be used when one objects wants to do something on another object destruction. Here we have a different situation: we have only 1 object: the widget. Subscribing to self has no meaning: the parent data is already destroyed when "destroyed" signal is emitted.